### ISSUE TITLE
Need a query that matches number and types of icons

### ISSUE THREAD

**Issue by ultimatefiend** [micahstairs/bga-innovation#205]
**NOTE:** Before you begin working on this, please **assign this issue to yourself**. If you do not do this, then there's the possibility that 2 people will end up doing the same task, which is not a good use of time.

This query is very specific to "Complex Numbers" which is a relic in the Artifacts expansion. (#184)

Also, this is for extra information on the query that is required and how they should match:

https://boardgamegeek.com/thread/1872362/complex-numbers-relic-what-has-match

**Comment by micahstairs** [comment]
Thanks for filing this. As we discussed over Discord, my idea is to encode each card's icons into an integer (or string) and store it in the card table. We can populate that column dynamically when the game is being setup. And then we can add something to be able to filter cards based on matching encoded values.

**Comment by lordmcfuzz** [comment]
I'm interested in this. will def have to be written and documented well so other expansions work change it.

**Comment by ultimatefiend** [comment]
Go for it.  :)  I assigned you.

**Comment by lordmcfuzz** [comment]
A quick thought: one way could be mathematical
icons are given their current 0-6 values -> add them up
bonus values given their values * 100 then added to the value above. 
Echo effects given a 10000 value and added to the value above
Plus signs given a 100000 value and added to the value above

this then could give you a finger print value that should be able to check for equality. bonus values multiplied by 100 keeps them out of the way of icon values and won't get into the 10k range where echo effects touch. The plus sign gets it's own digit. to be with.

There might some edgecase with cities and their weird 6 icon space existence.

**Comment by micahstairs** [comment]
If we are calculating this at the time that the table is setup, then we will have the freedom to change the formula when we add other expansions.

**Comment by micahstairs** [comment]
For the icons (1-6), you could assign each of them a prime number (2, 3, 5, 7, 11). And then you could take the product of them to build a unique integer. For example, if you have two icon 3's and 1 icon 4 then you'd get 5*5*7=175.

We could re-evaluate the formula once we have bonus values and plus signs to worry about.

Echo effects are supposed to be ignored according to the game designer.

**Comment by micahstairs** [comment]
Also, this might be a helpful commit to look at when adding support for a new option in the options array: https://github.com/micahstairs/bga-innovation/commit/f25678081e9714e06009986fcf4c4df36ed209e0.

In our case we will want to name the option something like `icons_hashcode` (I'm open to better names) and we'd set it to the hashcode of the card we are trying to match the icons with.

**Comment by lordmcfuzz** [comment]
Oh this could be a column on card and it would be populated by a trigger on card table insert.

The trigger would multiply the spot_# value by a not small prime number, say 13 or higher and then multiply the icons together to get the hash. This would remove the need for 6 conditionals or changing the spot_#/icon numbering scheme. 
 
 So a 1,1,1 card would be 2197 compared to a 1,2,e card being a 338 card.
 
 Echos are "ignored" but I don't think one could match bulb, bulb, echo, with anything other than a bulb, bulb, echo card.

**Comment by micahstairs** [comment]
Instead of using a trigger, it might be better to just run some SQL update queries in `setupNewGame`, so that it's in the same spot as:

```
// Add information to the database about which cards have a demand.
foreach ($this->textual_card_infos as $id => $card_info) {
  if ($card_info['i_demand_effect_1'] !== null) {
    self::DbQuery(self::format("UPDATE card SET has_demand = TRUE WHERE id = {id}", array("id" => $id)));
  }
}
```

Regarding echo effects, I think they shouldn't contribute anything to the formula at all, the same way we'd want to treat hexagon icons, and inspire effects. It's way too risky to include them into the computation, in case it makes something not match when it's actually supposed to. I can easily imagine there being a card with "bulb, bulb, echo" and another card being "bulb, bulb, inspire".

Can you elaborate on what you mean by 6 conditions? Why can't use use small primes?

**Comment by lordmcfuzz** [comment]
is a card with "bulb bulb echo" the same as "bulb bulb inspire"? -> I might be treating the echo/inspire as an icon in itself like a bonus but checking for existence rather than contents.

6 conditionals -> a switch case if doing it in JS. 

A sql trigger is faster and reduces db calls than a foreach. 

My thought on the double digit prime with multiplication is that it safely separates everything. If there is a fear of the number getting too large(2^31 large), then we can likely drop it.

potential cases using prime 13*spot, multiplying each spot together, e = echo, i = inspire, echo and inspire are ignored:
1,1,1 = 2197
6,e,i = 78
2,2,e = 676
1,1,3 = 6591

same as above but with 5 being the prime:
1,1,1 = 125
6,e,i = 30
2,2,e = 100
1,1,3 = 375

I could build up a combination spreadsheet to prove uniqueness. 1-6 for icons, ignore e and i. how we represent bonus/plus sign is a little important but not game breaking.

**Comment by micahstairs** [comment]
Take a look at https://boardgamegeek.com/thread/1872362/complex-numbers-relic-what-has-match. From the game designer: "Complex Numbers cares about both basic and special icons, and the matches must be exact. A 2 bonus is different from a 3 bonus. A search icon is different from a normal icon. However, echo effects and inspire effects are not icons."

This means we have the following icons: 6 basic resources, 6 resources with search icon, 1 plus sign, 11 bonus values. This is a total of 24 different possible values.

The reason I'm suggesting using the product of primes to represent this is because it's guaranteed to be unique. See https://www.varsitytutors.com/hotmath/hotmath_help/topics/prime-factorization#:~:text=The%20Fundamental%20Theorem%20of%20Arithmetic,prime%20factorization%20of%20the%20number.

Now the 24th prime number is 89 and we need to represent up to 6 icons, so that means the maximum value this would generate is 89^6=496,981,290,961. Is this value too big for SQL/PHP?

**Comment by micahstairs** [comment]
If that's too big, then we need to find a more efficient way to encode the numbers, and we will need to be more careful about collisions.

**Comment by micahstairs** [comment]
In any case, we have a lot of options for possible formulas, and after you implement a formula we can run a SQL query on the database to print out the list of cards which match with each other, and then we could manually sanity check them to see if any are supposed to be inequal.

**Comment by lordmcfuzz** [comment]
we should stay within signed int32 which is 2,147,483,648, and 89^6 is over that. 
I'll write something up.

**Comment by micahstairs** [comment]
Staying within int32 sounds like a good idea. Thanks!

**Comment by ultimatefiend** [comment]
Note that you have less than 32 bits to work with in the database entries.  Its like 21 or 22 bits.  I'd need to check the math to be sure.

**Comment by ultimatefiend** [comment]
0x8000000 is the maximum value (and its signed)

**Comment by lordmcfuzz** [comment]
I explored a pure sql computed column but I'm running into math issues with comparisons of [3,6,6]vs[5,5,5] and [1,4,4]vs[4,2,2]

I'm moving to the php for each route.

**Comment by lordmcfuzz** [comment]
draft-ish commit pushed for comment.

@ultimatefiend are you looking for a function that would do the comparison of two cards? what is your expected input and output?

**Comment by micahstairs** [comment]
I left some comments. You did the bulk of the work we need! The next step is just wiring it up an array-based `icon_hashes` option into the options array so that we can support "You may reveal a card from your hand having exactly the same icons, in type and number, as a top card on your board".

This will allow us to restrict which cards the player can select, to the set of cards in their hand whose icon_hash is equal to one of up to 5 icon hashes. Normally, we'd encode a small array of numbers into a single integer (since we need to store this in the DB), however, in this case each hash will need to be stored in a separate integer. So we will need to define 5 global variables (`icon_hash_1`, etc.).

We do something very comparable in https://github.com/micahstairs/bga-innovation/commit/6a92fd2dbdcc99b6f6403d70fe8e95f34e79dce5, which restricts the cards you can select based on their  `id` column. Sorry for the noisy diff, since I did some refactoring in that commit too.

**Comment by micahstairs** [comment]
Yeah I'm happy to discuss further and provide pointers, especially since this codebase is so new to you.

**Comment by lordmcfuzz** [comment]
So for the 
> "You may reveal a card from your hand having exactly the same icons, in type and number, as a top card on your board".
It seems like the work I did covers that. If you are comparing Pottery[id:0,  icon_hash:27] from your hand, you would find that Agriculture[id:9, icon_hash:27] as a top card would be a valid target when comparing the icon_hash.

The getCardInfo function currently already grabs * from the card, which includes the hash, so when looking at cards that are at the top of their respective piles, the icon_hash would be a property.

I'm unsure why we would need to define 5 global hash ids when we have one hash that is the key hash to compare against. Kind of like ~innovation.php:4716 in function selectEligibleCards() -> "//Condition for icon" code. If we have a condition that requires a comparison of icon_hash, I could see modeling it after that. One global variable set to the icon hash we are looking for, and then making an addition edit to dbquery blocks.

**Comment by micahstairs** [comment]
Yes, it's true that the cards themselves contain all of the information required for the comparison, but take a closer look at selectEligibleCards(). In order for it to know which cards the player is allowed to select, it does a SQL query, and it uses information stored in global variables to know which properties to search for.

In our case, we know that the cards we are allowed to select (from our hand) have to equal one up to 5 different hashes (one for each top card on our board). So in order to select the eligible cards, we need to have each of these 5 hashes stored in a global variable.

Does that make more sense?

**Comment by lordmcfuzz** [comment]
I have a better understanding. I guess I was looking at it backwards. I'll mess around and do some testing of it.

**Comment by micahstairs** [comment]
Great, good luck!

**Comment by micahstairs** [comment]
Re-opening this because we have the hash being computed, but we didn't finish the work described in the comments above to unblock the implementation of Complex Numbers.

**Comment by micahstairs** [comment]
I'm taking over the last piece of work for this.
